### PR TITLE
feat: use sepBy1Indent for tactic and conv blocks

### PR DESCRIPTION
**Pull request by gebner**
This is the first step towards using `colEq` in tactic/conv blocks, addressing #1606.

The main user-visible change is that `by simp rfl` (without newline or semicolon) is no longer supported.

### CODE REVIEW

**Comment by gebner** [comment]
I had to do a slightly ugly bootstrapping hack in 4657b744e00f645d0618ea539c02498acccd8ea7.

The issue is that parser aliases (such as aliases) are always taken from the previous stage (because they're defined using `builtin_initialize`).  With other syntax, you can use `parseQuotWithCurrentStage` to use the current stage.  But from what I can tell this does not affect parser aliases.

File: tests/lean/traceTacticSteps.lean.expected.out [hunk 1]
```diff
@@ -5,30 +5,30 @@
...
   [Elab.step.result] ?m
 [Elab.step] skip
 [Elab.step] 
       skip
       trivial
+      
 [Elab.step] 
       skip
       trivial
+
```
Comment (digama0) [inline tests/lean/traceTacticSteps.lean.expected.out:18]: It seems the formatter regressed here?

Comment (gebner) [reply tests/lean/traceTacticSteps.lean.expected.out:18]: This is fun.  It turns out that `tacticSeq` now parses the trailing newline (after `trivial`) as a separator.  And the formatter then prints this final newline-separator.  We might want to change the formatter to drop the final separator if it's a newline.

**Comment by gebner** [comment]
After this PR, it will be a very small change to enforce uniform start columns in tactic blocks (and `sepByIndent` generally): https://github.com/gebner/lean4/compare/sepbyindenttac...gebner:lean4:coleq?expand=1

File: src/Lean/Elab/Tactic/BuiltinTactic.lean [hunk 2]
```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
 @[builtinTactic paren] def evalParen : Tactic := fun stx =>
   evalTactic stx[1]
 
-def isCheckpointableTactic (arg : Syntax) : TacticM Bool := do
+def isCheckpointableTactic (arg : Syntax.Tactic) : TacticM Bool := do
```
Comment (leodemoura) [inline src/Lean/Elab/Tactic/BuiltinTactic.lean:38]: Why did you have to change to `Syntax.Tactic`?

File: src/Lean/Elab/Tactic/BuiltinTactic.lean [hunk 3]
```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
...
 
-  go (i : Nat) (acc : Array Syntax) (result : Array Syntax) : TacticM (Array Syntax) := do
+  go (i : Nat) (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : TacticM (Array Syntax.Tactic) := do
     if h : i < args.size then
       let arg := args.get ⟨i, h⟩
-      if (← isCheckpointableTactic arg[0]) then
+      if (← isCheckpointableTactic arg) then
         -- `argNew` is `arg` as singleton sequence. The idea is to make sure it does not satisfy `isCheckpointableTactic` anymore
-        let argNew := arg.setArg 0 (mkNullNode #[arg[0]])
+        let argNew := ⟨mkNullNode #[arg]⟩ -- HACK
```
Comment (leodemoura) [inline src/Lean/Elab/Tactic/BuiltinTactic.lean:62]: Could you please explain the "HACK"?

File: src/Lean/Elab/Tactic/BuiltinTactic.lean [hunk 4]
```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
...
-  push (acc : Array Syntax) (result : Array Syntax) : Array Syntax :=
+  push (acc : Array Syntax.Tactic) (result : Array Syntax.Tactic) : Array Syntax.Tactic :=
     if acc.isEmpty then
       result
     else
-      let ref := acc.back
-      let accSeq := mkNode ``Lean.Parser.Tactic.tacticSeq1Indented #[mkNullNode acc]
-      let checkpoint := mkNode ``Lean.Parser.Tactic.checkpoint #[mkAtomFrom ref "checkpoint", accSeq]
-      result.push (mkNode groupKind #[checkpoint, mkNullNode])
+      result.push <| Unhygienic.run do withRef acc.back do
```
Comment (leodemoura) [inline src/Lean/Elab/Tactic/BuiltinTactic.lean:54]: It is not clear to me why this change is needed.

File: src/Lean/Elab/Tactic/BuiltinTactic.lean [hunk 5]
```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
...
         let acc := acc.push argNew
         go (i+1) #[] (push acc result)
       else
         go (i+1) (acc.push arg) result
     else
       return result ++ acc
 
-/-- Evaluate `many (group (tactic >> optional ";")) -/
-def evalManyTacticOptSemi (stx : Syntax) : TacticM Unit := do
-  for seqElem in (← addCheckpoints stx.getArgs) do
```
Comment (leodemoura) [inline src/Lean/Elab/Tactic/BuiltinTactic.lean:74]: You are converting here to `Array Syntax.Tactic`. It is not clear to me why.

File: src/Lean/Elab/Tactic/BuiltinTactic.lean [hunk 6]
```diff
@@ -35,55 +35,52 @@ open Parser.Tactic
...
-    evalTactic seqElem[0]
-    saveTacticInfoForToken seqElem[1] -- add TacticInfo node for `;`
```
Comment (leodemoura) [inline src/Lean/Elab/Tactic/BuiltinTactic.lean:76]: It is not clear to me why we don't need this anymore.

Comment (leodemoura) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:76]: "this" here is the line `saveTacticInfoForToken seqElem[1]`

Comment (gebner) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:76]: Yes, I understand.  I thought this was subsumed by Sebastian's other tactic position tweaks.  And the tests passed (including the ones checking tactic states at semicolons), so this was plausible.

But it turns out this was indeed necessary, but only for the space after the semicolon...  I'm adding a test.
```
  tac1; tac2
     --^
```

File: src/Lean/Elab/Tactic/BuiltinTactic.lean (hunk 3 above)
Comment (gebner) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:62]: It's a hack because it creates invalid syntax (see pretty-printer issues in the Zulip thread):

```
import Lean open Lean Elab Command PrettyPrinter

elab "run" : command => do
  let tac ← `(tactic| simp)
  let tac : TSyntax `tactic := ⟨mkNullNode #[tac]⟩ -- HACK: produces invalid tactic
  let term ← `(by $tac:tactic)
  logInfo (← liftCoreM do ppTerm term)

run -- unknown constant 'null'
```

File: src/Lean/Elab/Tactic/BuiltinTactic.lean (hunk 6 above)
Comment (leodemoura) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:76]: Adding more tests is great. I am frequently worried about changes that affect the infoview because of the lack of tests.

File: src/Lean/Elab/Tactic/BuiltinTactic.lean (hunk 4 above)
Comment (gebner) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean]: Reverted.

File: tests/lean/traceTacticSteps.lean.expected.out (hunk 1 above)
Comment (gebner) [reply tests/lean/traceTacticSteps.lean.expected.out]: Final newline tweak implemented.

File: src/Lean/Elab/Tactic/BuiltinTactic.lean (hunk 5 above)
Comment (gebner) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:74]: (Note: I have rewritten this function one more time since your comment.)

The general reason is that the AST of tactic sequences has changed from `many1 (group (tactic >> optional ";"))` to `sepBy1Indent tactic`, i.e. from `[(group (skip) [])]` to `[(skip) []]`.  Having explicit types available (`Array Syntax.Tactic` instead of `Syntax`) makes the refactoring easier because the type system tells you what is a list of tactics, what is a tactic, etc., and you don't have to guess yourself what the `Syntax` means.

At first it seemed like it was not necessary to keep the separators, making `Array Syntax.Tactic` a natural choice.

I have now changed it back to untyped `Syntax` because we need to preserve the separators.  Ideally we'd have a `TSepIndentArray` for this because the manual index calculations are not particularly pretty.  (However that doesn't solve all problems either.  Even the `TSepArray` API is pretty bare-bones---and it doesn't preserve the separators either.)

Comment (gebner) [reply src/Lean/Elab/Tactic/BuiltinTactic.lean:74]: > `TSepIndentArray`

Oh I completely forgot we had this already.  It's called ``TSepArray `tactic ""``.

**Comment by gebner** [comment]
Re: saving tactic info at separators.

This prevents the regression with `tac1; tac2` noted above, but it is extremely fragile because the separators can be easily lost:
```lean
syntax "mytac" tacticSeq : tactic
macro_rules
  | `(tactic| mytac $[$tacs]*) => `(tactic| { $tacs* })

example : True ∨ True := by
  mytac constructor; constructor
-- still has the issue with the space after the semicolon
```

We probably need to fix this somewhere else.

**Comment by leodemoura** [comment]
Thanks for the changes. I will go over the code tomorrow, and try to merge it. If I am still confused, we can discuss during the dev meeting.
